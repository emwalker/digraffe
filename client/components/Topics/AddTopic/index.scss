.container {
  input, button {
    display: inline-block;
  }

  .topic-name {
    width: 75%;
    margin-right: 1em;
  }
}
