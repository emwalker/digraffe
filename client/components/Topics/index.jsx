// @flow
import React from 'react'
import { compose, isNil, map, prop, propOr, reject } from 'ramda'
import { graphql, createFragmentContainer } from 'react-relay'

import Topic from './Topic'
import AddTopic from './AddTopic'

const topicList = compose(reject(isNil), map(prop('node')), propOr([], 'edges'))

type Props = {
  viewer: {
    name: string,
  },
  organization: {
    topics: Object,
  },
  relay: {
    environment: Object,
  }
}

const Topics = ({ viewer: { name }, organization, relay }: Props) => (
  <div>
    <h1>Topics</h1>
    <div className="row">
      <div className="col">
        <p className="lead">
          List of topics visible to { name }
        </p>
      </div>
      <div className="col-6">
        <AddTopic
          className="test-add-topic"
          organization={organization}
          relay={relay}
        />
      </div>
    </div>
    <ul>
      {topicList(organization.topics).map(topic => (
        <Topic key={topic.id} topic={topic} />
      ))}
    </ul>
  </div>
)

export default createFragmentContainer(Topics, graphql`
  fragment Topics_viewer on User {
    name
  }

  fragment Topics_organization on Organization {
    id
    resourceId

    topics(first: 100) @connection(key: "Organization_topics") {
      edges {
        node {
          id
          name
          resourceId
          description
        }
      }
    }
  }
`)
