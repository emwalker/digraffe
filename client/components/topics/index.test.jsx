import React from 'react'
import { shallow } from 'enzyme'
import Topics from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

describe('<Topics />', () => {
  const viewer = {
    name: 'Rezrov',
  }

  const organization = {
    resourceId: 'organization:tyrell',
    topics: {
      edges: [
        {
          node: {
            name: 'Frotz',
            resourceId: '/topics/1234',
            id: '1234',
          },
        },
      ],
    },
  }

  const relay = {
    environment: {},
  }

  const wrapper = shallow(
    <Topics
      organization={organization}
      viewer={viewer}
      relay={relay}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('includes a form to add a topic', () => {
    expect(wrapper.find('.test-add-topic')).toHaveLength(1)
  })
})
