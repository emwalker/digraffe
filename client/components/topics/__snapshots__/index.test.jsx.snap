// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topics /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Topics
    organization={
      Object {
        "resourceId": "organization:tyrell",
        "topics": Object {
          "edges": Array [
            Object {
              "node": Object {
                "id": "1234",
                "name": "Frotz",
                "resourceId": "/topics/1234",
              },
            },
          ],
        },
      }
    }
    relay={
      Object {
        "environment": Object {},
      }
    }
    viewer={
      Object {
        "name": "Rezrov",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Topics
        </h1>,
        <div
          className="row"
        >
          <div
            className="col"
          >
            <p
              className="lead"
            >
              List of topics visible to 
              Rezrov
            </p>
          </div>
          <div
            className="col-6"
          >
            <AddTopic
              className="test-add-topic"
              organization={
                Object {
                  "resourceId": "organization:tyrell",
                  "topics": Object {
                    "edges": Array [
                      Object {
                        "node": Object {
                          "id": "1234",
                          "name": "Frotz",
                          "resourceId": "/topics/1234",
                        },
                      },
                    ],
                  },
                }
              }
              relay={
                Object {
                  "environment": Object {},
                }
              }
            />
          </div>
        </div>,
        <ul>
          <Unknown
            topic={
              Object {
                "id": "1234",
                "name": "Frotz",
                "resourceId": "/topics/1234",
              }
            }
          />
        </ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Topics",
        },
        "ref": null,
        "rendered": "Topics",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col"
            >
              <p
                className="lead"
              >
                List of topics visible to 
                Rezrov
              </p>
            </div>,
            <div
              className="col-6"
            >
              <AddTopic
                className="test-add-topic"
                organization={
                  Object {
                    "resourceId": "organization:tyrell",
                    "topics": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "id": "1234",
                            "name": "Frotz",
                            "resourceId": "/topics/1234",
                          },
                        },
                      ],
                    },
                  }
                }
                relay={
                  Object {
                    "environment": Object {},
                  }
                }
              />
            </div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p
                className="lead"
              >
                List of topics visible to 
                Rezrov
              </p>,
              "className": "col",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "List of topics visible to ",
                  "Rezrov",
                ],
                "className": "lead",
              },
              "ref": null,
              "rendered": Array [
                "List of topics visible to ",
                "Rezrov",
              ],
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <AddTopic
                className="test-add-topic"
                organization={
                  Object {
                    "resourceId": "organization:tyrell",
                    "topics": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "id": "1234",
                            "name": "Frotz",
                            "resourceId": "/topics/1234",
                          },
                        },
                      ],
                    },
                  }
                }
                relay={
                  Object {
                    "environment": Object {},
                  }
                }
              />,
              "className": "col-6",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "test-add-topic",
                "organization": Object {
                  "resourceId": "organization:tyrell",
                  "topics": Object {
                    "edges": Array [
                      Object {
                        "node": Object {
                          "id": "1234",
                          "name": "Frotz",
                          "resourceId": "/topics/1234",
                        },
                      },
                    ],
                  },
                },
                "relay": Object {
                  "environment": Object {},
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown
              topic={
                Object {
                  "id": "1234",
                  "name": "Frotz",
                  "resourceId": "/topics/1234",
                }
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1234",
            "nodeType": "function",
            "props": Object {
              "topic": Object {
                "id": "1234",
                "name": "Frotz",
                "resourceId": "/topics/1234",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Topics
          </h1>,
          <div
            className="row"
          >
            <div
              className="col"
            >
              <p
                className="lead"
              >
                List of topics visible to 
                Rezrov
              </p>
            </div>
            <div
              className="col-6"
            >
              <AddTopic
                className="test-add-topic"
                organization={
                  Object {
                    "resourceId": "organization:tyrell",
                    "topics": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "id": "1234",
                            "name": "Frotz",
                            "resourceId": "/topics/1234",
                          },
                        },
                      ],
                    },
                  }
                }
                relay={
                  Object {
                    "environment": Object {},
                  }
                }
              />
            </div>
          </div>,
          <ul>
            <Unknown
              topic={
                Object {
                  "id": "1234",
                  "name": "Frotz",
                  "resourceId": "/topics/1234",
                }
              }
            />
          </ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Topics",
          },
          "ref": null,
          "rendered": "Topics",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col"
              >
                <p
                  className="lead"
                >
                  List of topics visible to 
                  Rezrov
                </p>
              </div>,
              <div
                className="col-6"
              >
                <AddTopic
                  className="test-add-topic"
                  organization={
                    Object {
                      "resourceId": "organization:tyrell",
                      "topics": Object {
                        "edges": Array [
                          Object {
                            "node": Object {
                              "id": "1234",
                              "name": "Frotz",
                              "resourceId": "/topics/1234",
                            },
                          },
                        ],
                      },
                    }
                  }
                  relay={
                    Object {
                      "environment": Object {},
                    }
                  }
                />
              </div>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p
                  className="lead"
                >
                  List of topics visible to 
                  Rezrov
                </p>,
                "className": "col",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "List of topics visible to ",
                    "Rezrov",
                  ],
                  "className": "lead",
                },
                "ref": null,
                "rendered": Array [
                  "List of topics visible to ",
                  "Rezrov",
                ],
                "type": "p",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <AddTopic
                  className="test-add-topic"
                  organization={
                    Object {
                      "resourceId": "organization:tyrell",
                      "topics": Object {
                        "edges": Array [
                          Object {
                            "node": Object {
                              "id": "1234",
                              "name": "Frotz",
                              "resourceId": "/topics/1234",
                            },
                          },
                        ],
                      },
                    }
                  }
                  relay={
                    Object {
                      "environment": Object {},
                    }
                  }
                />,
                "className": "col-6",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "className": "test-add-topic",
                  "organization": Object {
                    "resourceId": "organization:tyrell",
                    "topics": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "id": "1234",
                            "name": "Frotz",
                            "resourceId": "/topics/1234",
                          },
                        },
                      ],
                    },
                  },
                  "relay": Object {
                    "environment": Object {},
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown
                topic={
                  Object {
                    "id": "1234",
                    "name": "Frotz",
                    "resourceId": "/topics/1234",
                  }
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1234",
              "nodeType": "function",
              "props": Object {
                "topic": Object {
                  "id": "1234",
                  "name": "Frotz",
                  "resourceId": "/topics/1234",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
